modules/kvformat/filterx-func-parse-kv.c

```diff
@@ -0,0 +1,294 @@
+
+enum parse_kv_opts
+{
+  PARSE_KV_OPTS_PREFIX = 0,
+  PARSE_KV_OPTS_VALUE_SEP = 1,
+  PARSE_KV_OPTS_PAIR_SEP = 2,
+  PARSE_KV_OPTS_STRAY_WORDS_VAL_NAME = 3,
+};
+
+static const gchar *parse_kv_opts_names[PARSE_KV_OPTS_COUNT] =
+{
+  "prefix",
+  "value_separator",
+  "pair_separator",
+  "stray_words_value_name",
+};
+
+typedef struct FilterXFunctionParseKV_
+{
+  FilterXFunction super;
+  FilterXExpr *msg;
+  LogParser *kv_parser;
+  GlobalConfig *cfg;
+} FilterXFunctionParseKV;
+
+static gboolean
+_set_json_vals_foreach(NVHandle handle, const gchar *name,
+                       const gchar *value, gssize value_len,
+                       LogMessageValueType type, gpointer user_data)
+{
+  msg_trace("KV_PAIR found",
+            evt_tag_str("type", log_msg_value_type_to_str(type)),
+            evt_tag_str("key", name),
+            evt_tag_str("value", value));
+
+  FilterXObject *obj = (FilterXObject *)user_data;
+  FilterXObject *key = filterx_string_new(name, -1);
+  FilterXObject *val = filterx_string_new(value, value_len);
+
+  gboolean ok = filterx_object_set_subscript(obj, key, val);
+
+  filterx_object_unref(key);
+  filterx_object_unref(val);
+
+  gboolean stop = !ok;
+  return stop;
+}
+
+static gboolean
+_apply_parse_kv_option(LogParser *kv_parser, int opt, const gchar *opt_val, GError **error)
+{
+  switch (opt)
+    {
+    case PARSE_KV_OPTS_PREFIX:
+      kv_parser_set_prefix(kv_parser, opt_val);
+      break;
+    case PARSE_KV_OPTS_VALUE_SEP:
+    {
+      if (strlen(opt_val) < 1)
+        {
+          g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                      "%s argument can not be empty. " FILTERX_FUNC_PARSE_KV_USAGE,
+                      parse_kv_opts_names[opt]);
+          return FALSE;
+        }
+      kv_parser_set_value_separator(kv_parser, opt_val[0]);
+      break;
+    }
+    case PARSE_KV_OPTS_PAIR_SEP:
+      kv_parser_set_pair_separator(kv_parser, opt_val);
+      break;
+    case PARSE_KV_OPTS_STRAY_WORDS_VAL_NAME:
+      kv_parser_set_stray_words_value_name(kv_parser, opt_val);
+      break;
+    default:
+      g_assert_not_reached();
+      break;
+    }
+  return TRUE;
+}
+
+static FilterXObject *
+_eval(FilterXExpr *s)
+{
+  FilterXFunctionParseKV *self = (FilterXFunctionParseKV *) s;
+
+  FilterXObject *obj = filterx_expr_eval(self->msg);
+  if (!obj)
+    return NULL;
+
+  gboolean ok = FALSE;
+
+  gsize len;
+  const gchar *str;
+  if (filterx_object_is_type(obj, &FILTERX_TYPE_NAME(string)))
+    str = filterx_string_get_value(obj, &len);
+  else if (filterx_object_is_type(obj, &FILTERX_TYPE_NAME(message_value)))
+    str = filterx_message_value_get_value(obj, &len);
+  else
+    goto exit;
+
+  LogMessage *msg = log_msg_new_empty();
+  FilterXObject *result = filterx_json_object_new_empty();
+
+  LogPathOptions path_options = LOG_PATH_OPTIONS_INIT; // can use NULL instead of this
+  ok = log_parser_process(self->kv_parser, &msg, &path_options, str, len);
```

Comment:
we should not use the kv-parser() functionality itself, rather we should use the lower level KVScanner instance, which would give us the same functionality but without tying us to LogMessage et al.